### PR TITLE
Add ExecutableDLQObserver

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added `queues.ExecutableDLQObserver` which adds logging and metrics to the history task DLQ.

**Why?**
In order for us to alert on things like high DLQ latency, high DLQ rate, or high DLQ failure rate.

**How did you test it?**
There's 100% test coverage for the new code. 

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/queues/executable_dlq_observer.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	} else if err == nil && o.terminalFailureTime != nil {
+		latency := o.timeSource.Now().Sub(*o.terminalFailureTime)
+		o.metricsHandler.Timer(metrics.TaskDLQSendLatency.GetMetricName()).Record(latency)
+		logger := o.getLogger()
+		logger.Info("Task sent to DLQ")
+	}
+	return err
+}
+
+func (o *ExecutableDLQObserver) getLogger() log.Logger {
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/queues/executable_dlq_observer.go:103]: Done, thanks!

File: service/history/queues/executable_dlq_observer.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		logger.Error("A terminal error occurred while processing this task", tag.Error(err))
+		now := o.timeSource.Now()
+		o.terminalFailureTime = &now
+	} else if errors.Is(err, ErrSendTaskToDLQ) {
+		o.metricsHandler.Counter(metrics.TaskDLQFailures.GetMetricName()).Record(1)
+		logger := o.getLogger()
+		logger.Error("Failed to send history task to the DLQ", tag.Error(err))
+	} else if err == nil && o.terminalFailureTime != nil {
+		latency := o.timeSource.Now().Sub(*o.terminalFailureTime)
+		o.metricsHandler.Timer(metrics.TaskDLQSendLatency.GetMetricName()).Record(latency)
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/queues/executable_dlq_observer.go]: Yeah, probably not that useful. I think I was trying to track the E2E time from detection until it's actually DLQ'd, but that doesn't seem as important as the latency of the actual work. I changed it to just record the latency of the final execution, which successfully sends the task to the DLQ.

File: service/history/queues/executable_dlq_observer.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		numHistoryShards:    numHistoryShards,
+		timeSource:          timeSource,
+		terminalFailureTime: nil,
+	}
+}
+
+func (o *ExecutableDLQObserver) Execute() error {
+	err := o.ExecutableDLQ.Execute()
+	if errors.Is(err, ErrTerminalTaskFailure) {
+		o.metricsHandler.Counter(metrics.TaskTerminalFailures.GetMetricName()).Record(1)
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/queues/executable_dlq_observer.go]: Added, and I added tests for them

File: service/history/dlq.go [hunk 4]
```diff
@@ -25,6 +25,10 @@
 package history
 
 import (
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/service/history/configs"
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/dlq.go]: Done. I also tried making a config change for GoLand here, but it doesn't seem to work:
image

File: service/history/queues/executable_dlq_observer.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	} else if err == nil && o.isDLQ {
+		latency := o.timeSource.Now().Sub(now)
+		o.getMetricsHandler().Timer(metrics.TaskDLQSendLatency.GetMetricName()).Record(latency)
+		logger := o.getLogger()
+		logger.Info("Task sent to DLQ")
+	}
+	return err
+}
+
+func (o *ExecutableDLQObserver) getMetricsHandler() metrics.Handler {
```
Comment (yycptt) [inline service/history/queues/executable_dlq_observer.go:90]: we just need to tag the metrics handler once and reuse the tagged handler I think. Same for the logger. 

I remember I've implemented a lazyLogger and used in `executableImpl`, you might want to take a look.

File: service/history/queues/executable_dlq_observer.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		numHistoryShards:  numHistoryShards,
+		namespaceRegistry: namespaceRegistry,
+		timeSource:        timeSource,
+		logger:            logger,
+		metricsHandler:    metricsHandler,
+		isDLQ:             false,
+	}
+}
+
+func (o *ExecutableDLQObserver) Execute() error {
```
Comment (yycptt) [inline service/history/queues/executable_dlq_observer.go:69]: Not related to this PR but it just occurred to me.

If the State of the underlying task is not Pending (e.g. queue is shutting down or multi-cursor impl want to offload/delete some tasks from memory) anymore, the execution should be skipped.
https://github.com/temporalio/temporal/blob/main/service/history/queues/executable.go#L188

**Review by yycptt — APPROVED** [review]

**Pull request merged**